gettsim/transfers/grundrente.py

```diff
@@ -0,0 +1,397 @@
+    out = (
+        ges_rentenv_params["höchstwert_grundrente"]["base"]
+        + ges_rentenv_params["höchstwert_grundrente"]["increment"] * months_above_thresh
+    )
+
+    return out
+
+
+@add_rounding_spec(params_key="ges_rentenv")
+def grundr_zuschlag_bonus_entgeltp(
+    _durchschnittl_entgeltp_g_r_bewertungszeiten: FloatSeries,
+    höchstwert_grundr_zuschlag_m: FloatSeries,
+    grundrentenzeiten: IntSeries,
+    ges_rentenv_params: dict,
+) -> FloatSeries:
+    """Calculate additional Entgeltpunkte for pensioner.
+
+    In general, the average of monthly Entgeltpunkte earnd in Grundrentenzeiten is
+    doubled, or extended to the individual Höchstwert if doubling would exceed the
+    Höchstwert. Then, the value is multiplied by 0.875.
+
+    Legal reference: § 76g SGB VI
+
+    Parameters
+    ----------
+    _durchschnittl_entgeltp_g_r_bewertungszeiten
+        See :func:`_durchschnittl_entgeltp_g_r_bewertungszeiten`.
+    höchstwert_grundr_zuschlag_m
+        See :func:`höchstwert_grundr_zuschlag_m`.
+    grundrentenzeiten
+        See basic input variable :ref:`grundrentenzeiten <grundrentenzeiten>`.
+    ges_rentenv_params
+        See params documentation :ref:`ges_rentenv_params <ges_rentenv_params>`.
+
+    Returns
+    -------
+
+    """
+    out = pd.Series(0, index=grundrentenzeiten.index)
+
+    # Case 1: Entgeltpunkte less than half of Höchstwert
+    below_half_höchstwert = _durchschnittl_entgeltp_g_r_bewertungszeiten <= (
+        0.5 * höchstwert_grundr_zuschlag_m
+    )
+    out.loc[below_half_höchstwert] = _durchschnittl_entgeltp_g_r_bewertungszeiten
+
+    # Case 2: Entgeltpunkte more than half of Höchstwert, but below Höchstwert
+    cond = ~below_half_höchstwert & (
+        _durchschnittl_entgeltp_g_r_bewertungszeiten < höchstwert_grundr_zuschlag_m
+    )
+    out.loc[cond] = (
+        höchstwert_grundr_zuschlag_m - _durchschnittl_entgeltp_g_r_bewertungszeiten
+    )
+
+    # Case 3: Entgeltpunkte above Höchstwert
+    cond = _durchschnittl_entgeltp_g_r_bewertungszeiten > höchstwert_grundr_zuschlag_m
+    out.loc[cond] = 0
+
+    # Set to 0 if Grundrentenzeiten below minimum
+    gr_zeiten_below_min = (
+        grundrentenzeiten < ges_rentenv_params["grundrentenzeiten"]["min"]
+    )
+    out.loc[gr_zeiten_below_min] = 0
+
+    # Multiply additional Engeltpunkte by factor
+    out = out * ges_rentenv_params["faktor_bonus_grundrente"]
+
+    return out
+
+
+def eink_excl_grundr_zuschlag_m_tu(
+    eink_excl_grundr_zuschlag_m: FloatSeries, tu_id: IntSeries
+) -> FloatSeries:
+    """Aggregate the income relevant for income crediting rule of
+    Grundrentenzuschlag on tax unit level.
+
+    Parameters
+    ----------
+    eink_excl_grundr_zuschlag_m
+        See :func:`eink_excl_grundr_zuschlag_m`.
+    tu_id
+        See basic input variable :ref:`tu_id <tu_id>`.
+
+    Returns
+    -------
+
+    """
+    return eink_excl_grundr_zuschlag_m.groupby(tu_id).sum()
+
+
+def eink_excl_grundr_zuschlag_m(
```

Comment:
No good test cases available as test cases usually just state "Einkommen X€". Also because we make a few assumptions here (because no information on last year income for most income sources).

Do you think it would be worth it to make up a (regression) test?

Reply:
Just realized that a few functions are affected by that. So I will create a few tests for it.

Reply:
Cool, thanks! 

Only looked at the docstring in detail now. 

We will need to come up with something better for 

`grundr_zuschlag_eink_m`

and 

`eink_excl_grundr_zuschlag_m`

This is too confusing for my simple mind. Probably best to talk that over at some point!

Reply:
Maybe just get rid of the second function?

Reply:
Not sure, can we? I thought that is past income determining current eligibility and level. And the first function determines the tapering given concurrent income? So both seem necessary?

Reply:
`eink_vor_grundr_zuschlag_m` (former `eink_excl_grundr_zuschlag_m`) is total income relevant for Grundrentenzuschlag (before considerung allowances). `vor_grundr` just means that Grundrentenzuschlag is not part of this income, but all other sources.

`grundr_zuschlag_eink_m` is income which is deducted from Grundrentenzuschlag (after considering allowances).

The first one is just used in the second one, so we could put them together. I would, however, prefer if we find a good way to rename them. (Talking about it might be a good idea)

Reply:
Thanks! Just tried to understand this better, a couple of notes:
- We should strive to include references in docstrings of functions like `eink_excl_grundr_zuschlag_m`. That is, not fully obvious why these particular five types of income are considered and no help resorting to the references included with parameters. In the current state, I would have no idea where all that comes from.
- I do not understand this to-do, certainly not obvious where pension income enters `freibeträge`? 

    #ToDo: Freibeträge for income are currently not considered
    #ToDO: as freibeträge_tu depends on pension income

- I continue to be confused by the `[transfer]_eink` naming, i.e., without anything like `eink_anr`. Should reconsider in the big renaming thing.
- For now, I would actually suggest merging the two functions. We can reconsider when re-writing `grundr_zuschlag_eink_m` as part of the move to a scalar-based solution. Always easier to add variables ex post than removing some.

Reply:
> We should strive to include references in docstrings of functions like eink_excl_grundr_zuschlag_m. That is, not fully obvious why these particular five types of income are considered and no help resorting to the references included with parameters. In the current state, I would have no idea where all that comes from.

Agreed!

> I do not understand this to-do, certainly not obvious where pension income enters freibeträge?

through `ges_krankenv_beitr_m` -> `vorsorge` -> `freibeträge`. I added this information to the ToDo

> I continue to be confused by the [transfer]_eink naming, i.e., without anything like eink_anr. Should reconsider in the big renaming thing.

Alright!

> For now, I would actually suggest merging the two functions. We can reconsider when re-writing grundr_zuschlag_eink_m as part of the move to a scalar-based solution. Always easier to add variables ex post than removing some.

Will do.

Reply:
Done now.

Also added new tests as discussed: code coverage of the new code is 100% now.

Reply:
Excellent! Then please check whether the documentation changes I suggested are correct!